PCbuild/get_external.py

```diff
@@ -1,36 +1,62 @@
 #!/usr/bin/env python3
 
 import argparse
+import contextlib
+import hashlib
+import io
+import json
 import os
 import pathlib
+import shutil
 import sys
 import time
 import urllib.error
 import urllib.request
 import zipfile
 
 
-def retrieve_with_retries(download_location, output_path, reporthook,
-                          max_retries=7):
-    """Download a file with exponential backoff retry and save to disk."""
+# Mapping of binary dependency tag to GitHub release asset ID
+TAG_TO_ASSET_ID = {
+    'llvm-20.1.8.0': 301710576,
+}
+
+
+def request_with_retry(request_func, *args, max_retries=7,
+                       err_msg='Request failed.', **kwargs):
+    """Make a request using request_func with exponential backoff"""
     for attempt in range(max_retries + 1):
         try:
-            resp = urllib.request.urlretrieve(
-                download_location,
-                output_path,
-                reporthook=reporthook,
-            )
+            resp = request_func(*args, **kwargs)
         except (urllib.error.URLError, ConnectionError) as ex:
             if attempt == max_retries:
-                msg = f"Download from {download_location} failed."
-                raise OSError(msg) from ex
+                raise OSError(err_msg) from ex
             time.sleep(2.25**attempt)
         else:
             return resp
 
 
-def fetch_zip(commit_hash, zip_dir, *, org='python', binary=False, verbose):
-    repo = f'cpython-{"bin" if binary else "source"}-deps'
+def retrieve_with_retries(download_location, output_path, reporthook):
+    """Download a file with retries."""
+    return request_with_retry(
+        urllib.request.urlretrieve,
+        download_location,
+        output_path,
+        reporthook,
+        err_msg=f'Download from {download_location} failed.',
+    )
+
+
+def get_with_retries(url, headers):
+    req = urllib.request.Request(url=url, headers=headers, method='GET')
+    return request_with_retry(
+        urllib.request.urlopen,
+        req,
+        err_msg=f'Request to {url} failed.'
+    )
```

Comment:
Why do we have two similar functions? I see they are different on the function they use. Why do we need them both?
Can we reduce the usage to either one of them?

Reply:
This is code I wrote, so I can try and answer this :)

The idea is that we need to do two types of requests:
1) download a file (retrieve_with_retries)
2) get some JSON from GitHub's API (get_with_retries)

For 1, `urlretrieve` makes things a lot simpler. If we just had one generic function here we'd end up re-implementing `urlretrieve` (to handle things like writing downloaded content to the output file, etc).

For 2, we need to pass `Accept` headers to a GET request against GitHub's API to get information from the API about a release asset, so urlretrieve isn't quite right, and we need to use a `urllib.request.Request` to customize headers.